atef/widgets/config/find_replace.py

```diff
@@ -0,0 +1,1009 @@
+    final_step = path[-1]
+    # need the item one step before the last to perform the assignment on
+    parent_item = get_item_from_path(path[:-1], item=item)
+
+    if final_step[0] == "__dictkey__":
+        parent_item[replace_fn(final_step[1])] = parent_item.pop(final_step[1])
+    elif final_step[0] in ("__dictvalue__", "__list__"):
+        # replace value
+        old_value = parent_item[final_step[1]]
+        parent_item[final_step[1]] = replace_fn(old_value)
+    elif final_step[0] == "__enum__":
+        parent_item = get_item_from_path(path[:-2], item=item)
+        old_enum: Enum = getattr(parent_item, path[-2][1])
+        new_enum = getattr(final_step[1], replace_fn(old_enum.name))
+        setattr(parent_item, path[-2][1], new_enum)
+    else:
+        # simple field paths don't have a final (__sth__, ?) segement
+        old_value = getattr(parent_item, path[-1][1])
+        setattr(parent_item, path[-1][1], replace_fn(old_value))
+
+
+def get_default_match_fn(search_regex: re.Pattern) -> Callable:
+    def match_fn(match):
+        return search_regex.search(str(match)) is not None
+
+    return match_fn
+
+
+def get_default_replace_fn(replace_text: str, search_regex: re.Pattern) -> Callable:
+    def replace_fn(value):
+        if isinstance(value, str):
+            return search_regex.sub(replace_text, value)
+        elif isinstance(value, int):
+            # cast to float first
+            return int(float(value))
+        else:  # try to cast as original type
+            return type(value)(replace_text)
+
+    return replace_fn
+
+
+def verify_changes(parent, orig_file: Union[ConfigurationFile, ProcedureFile]) -> None:
+    # check to make sure changes are valid
+    try:
+        if isinstance(orig_file, ConfigurationFile):
+            _ = PreparedFile.from_config(orig_file)
+        elif isinstance(orig_file, ProcedureFile):
+            # clear all results when making a new run tree
+            _ = PreparedProcedureFile.from_origin(orig_file)
+        else:
+            QtWidgets.QMessageBox.warning(
+                parent,
+                'Verification FAIL',
+                'File type not recognized.'
+            )
+            return
+    except Exception as ex:
+        logger.debug(ex)
+        QtWidgets.QMessageBox.warning(
+            parent,
+            'Verification FAIL',
+            'File could not be prepared successfully, edits will not work'
+        )
+        return
+
+    QtWidgets.QMessageBox.information(
+        parent,
+        'Verification PASS',
+        'File prepared successfully, edits should work'
+    )
+
+
+@dataclass
+class FindReplaceAction:
+    target: Union[ConfigurationFile, ProcedureFile]
+    path: List[Tuple[Any, Any]]
+    replace_fn: Callable
+
+    def apply(
+        self,
+        target: Optional[Union[ConfigurationFile, ProcedureFile]] = None,
+        path: Optional[List[Tuple[Any, Any]]] = None,
+        replace_fn: Optional[Callable] = None
```

Comment:
```python
ReplaceFunction = Callable[...something here...]
```
and
```suggestion
        replace_fn: Optional[ReplaceFunction] = None
```

Would recommend typing the "replace_fn" so you can tell if it's valid